Minor additions to taint propagators (#687)
returntocorp/semgrep-docs: docs/experiments/taint-propagators.md

# Taint propagators

> Taint propagators allow you to customize how taint is propagated. TODO

Author: @IagoAbal I made this PR so we can bring some of the ideas from the previous discussion here. Please, can we say a bit more here to specify taint propagators? We somehow specify them in action below, but I am still not sure if we state the essence. Maybe @efortuna will have an idea here also?

Reviewer: yeah, this faces the problem of a circular definition: "don't define something with the same words as the actual thing" 

If you need inspiration, you can always look at page 5 in my script: https://docs.google.com/document/d/1hqB-a_wkqrEwk_ZsFnlZKOFo5HPeUdsGPMRj8umTDxk/edit

```suggestion
Taint propagators allow you to specify additional structures through which taint spreads, that Semgrep by default might not mark as tainted.
```

> Taint propagators allow you to specify additional structures through which taint spreads, that Semgrep by default might not mark as tainted.

Author: ```suggestion
The default configuration of taint mode in Semgrep sometimes misses specific tainted structures. Taint propagators allow you to specify additional structures through which taint spreads. This results in taint mode rules with taint propagators that are more efficient in detecting tainted structures. 
```

> The default configuration of taint mode in Semgrep sometimes misses specific tainted structures. Taint propagators allow you to specify additional structures through which taint spreads. This results in taint mode rules with taint propagators that are more efficient in detecting tainted structures. 

Author: ```suggestion
The default configuration of taint mode in Semgrep sometimes misses specific tainted structures. Taint propagators allow you to specify additional structures through which taint propagates. This results in taint mode rules with taint propagators that are more efficient in detecting tainted structures. 
```

> The default configuration of taint mode in Semgrep sometimes misses specific tainted structures. Taint propagators allow you to specify additional structures through which taint propagates. This results in taint mode rules with taint propagators that are more efficient in detecting tainted structures. 

Author: ```suggestion
The default configuration of taint mode in Semgrep sometimes misses specific tainted structures in the code. Taint propagators allow you to specify additional structures through which taint propagates. This results in taint mode rules with taint propagators that are more efficient in detecting tainted structures. 
```

> The default configuration of taint mode in Semgrep sometimes misses specific tainted structures. Taint propagators allow you to specify additional structures through which taint propagates. This results in taint mode rules with taint propagators that are more efficient in detecting tainted structures. 

Author: ```suggestion
The default configuration of taint mode in Semgrep sometimes misses specific tainted structures. Taint propagators allow you to specify additional structures through which taint propagates. This enables you to create and use taint mode rules with taint propagators that are more efficient in detecting tainted structures. 
```

```

In the example above, whenever Semgrep finds the pattern `$X.add($Y)`, it checks if
`$Y` is tainted. If `$Y` is tainted, Semgrep propagates that taint to `$X`. 
Thus, adding tainted data to a set marks the set itself as tainted.

<iframe src="https://semgrep.dev/embed/editor?snippet=7lNe" border="0" frameBorder="0" width="100%" height="432"></iframe>

## Requirements of taint propagator rules

A taint propagator rules necessitate:

> - Any arbitrary pattern that must bind two metavariables of your choice.

Reviewer: what is this referring to?

Author: @efortuna looking at our example rule, we have - `- pattern: $X.add($Y)` - what do you think about the rewrite below?

```suggestion
- Rule pattern that binds two metavariables. For example: `- pattern: $X.add($Y)`
```

Reviewer: Adam your suggestion doesn't fit with the line above "A taint propagator rule necessitates (or must contain)"

but my question was more like, what are these two metavariables? What is their function? why are we binding them? it's really vague and I don't understand what it's talking about.

Author: @efortuna you are right. I cannot guess this information. @IagoAbal please, can you help us out here? (See the comment above.)

Author: ```suggestion
- Any arbitrary pattern that must bind two metavariables of your choice. For example `- pattern: $X.add($Y)`. Then specify which falls under `from` and which falls under `to`.
```

Reviewer: Y'all, my point in my whole line of questioning was pointing out we literally don't say what to do with this arbitrary pattern or what it's for (and "arbitrary" is odd here. All patterns specified in a semgrep rule are "arbitrary"...). As written, this arbitrary pattern is just floating in space. The point is it is connected to the `pattern` key, and this pattern indicates where taint is propagated!

```suggestion
- A `pattern` containing two metavariables, describing where taint is propagated _to_ and _from_.
```

Author: ```suggestion
- A `pattern` containing **two** metavariables. These two metavariables specify where taint is propagated **from** and **to**. For example, pattern `$X.add($Y)` includes two metavariables `$X` and `$Y`. Then specify which metavariable falls under `from` and which falls under `to` (see more information below).
```

> - You must specify the `from` and the `to` metavariables. TODO

Author: @IagoAbal can we say something about `from` and the `to` metavariables? Is this specifying the "span" of variables that Semgrep should look for?

Reviewer: Some details on [page 5 of my script](https://docs.google.com/document/d/1hqB-a_wkqrEwk_ZsFnlZKOFo5HPeUdsGPMRj8umTDxk/edit). 
```suggestion
- You must specify the `from` and the `to` metavariables. TODO
-  The `from` and `to` metavariables are required. `from` indicates where the tainted source must to pass “through” to propagate the taint. `to` specified what additional object or structure is tainted as a result.
```
@adamkvitek I'm sure you can clean up that suggestion to fit the tech writer standards, but that's a first stab.

> - The `from` and `to` metavariables are required. `from` indicates where the tainted source must to pass “through” to propagate the taint. `to` specified what additional object or structure is tainted as a result.

Author: ```suggestion
- The `from` and `to` metavariables. The `from` indicates the starting variable from which the tainted source passes through to propagate the taint. The `to` specifies what additional object or structure is tainted as a result.
```

> - The `from` and `to` metavariables. The `from` indicates the starting variable from which the tainted source passes through to propagate the taint. The `to` specifies what additional object or structure is tainted as a result.

Reviewer 2: Actually neither `from` nor `to` are restricted to match variables, and `from` in particular may not be a variable. The `to` will more often be a variable.

Author: @IagoAbal so what does the `from` specify mostly? Feel free to suggest your wording of the sentence, as it will reduce the guess work.

Author: I mean, this answer needs more specification. @IagoAbal

Reviewer: @IagoAbal why didn't you say this when reviewing my video script??

If this is the case, what are situations where people are using it not involving variables? Especially `from`? Why would someone use it that way, and how would you describe it?

As @adamkvitek points out, not adding any information other than saying "that's incorrect" increases the amount of back and forth on these sorts of reviews, which was brought up in the meeting yesterday as being stressful.

Reviewer 2: Initially it was restricted to variables, then it was generalized. In general `from` and `to` can match any expression, but being variables should be the most common case, specially for `to`. The feature is new and experimental, so we don't know how people will be using it, but I saw the use case of tainting a data structure several times, hence I added the feature. The video is not incorrect!

Author: @efortuna @IagoAbal thank you! I am suggesting edits below - please let me know if this is correct.

Should I substitute the word "structure" with "data structure" to be consistent (it is used in the introduction also)?

```suggestion
- The `from` and `to` metavariables.
    - The `from` indicates the starting variable or expression from which the tainted source passes through to propagate the taint. 
    - The `to` specifies what additional object or data structure, most commonly a variable, is tainted as a result.
```

Reviewer: @IagoAbal can you provide a suggestion for wording so that someone coming to this for the first time would know how to conceptualize `from` and `to` if they can be anything?

Reviewer 2: The use of the term "taint source" here is a bit confusing to me, because a "taint source" has a special meaning in taint-mode. It's not the "tainted source" that passes through the propagator, but the "taint". The "taint source" is the origin of the taint, e.g. `dangerous_user_input`.

You may want to rewrite it but this would be my proposal:

> The `from` metavariable specifies the _entry point_ of the propagator, it should match a variable, or generally an expression. The `to` metavariable specifies the _exit point_, it should also match a variable (or generally an expression), and typically denotes an object or data structure. Taint that reaches the entry point (the `from`) will be propagated to the exit point (the `to`) by the taint engine.

Author: Thanks, @IagoAbal, for this great information.

By " it should match a variable" do you mean the `from` metavariable:

A. Matches a variable or expression.
B. A user should specify it as a variable or expression.
C. Both A. and B. (I guess you mean this one.) When a user specifies it as a variable or expression - Semgrep matches variables or expressions with it. I will use the imperative word "specify" below, so the reader is pushed to action here. Let me know if it is correct.

Working with your suggestion:

```suggestion
- The `from` and `to` metavariables.
    - The `from` metavariable specifies the entry point of the taint propagator. Specify a value of `from` metavariable as a variable, or generally as an expression. 
    - The `to` metavariable specifies the exit point of the taint propagator. Specify a value of `to` metavariable as a variable, or generally as an expression. This metavariable typically denotes an object or data structure. Taint that reaches the entry point (the `from`)  is propagated to the exit point (the `to`) by the taint engine.
```

Reviewer: I don't think we need the an additionally indented list. 

```suggestion
- A `from` metavariable specifying the entry point of the taint. 
- A `to` metavariable specifying where the tainted data is propagated to, typically an object or data structure.

`to` and `from` can contain either a variable or expression.
```

Side note, if we're being pedantic: @IagoAbal You say "I saw the use case of tainting a data structure several times" but a data structure itself is not an expression. You saw people taint a data structure _in_ a `from` or `to` value?

Author: ```suggestion
- Use `to` and `from` metavariables. Both of these metavariables must contain either a **variable** or an **expression**.
    - The `from` metavariable specifies the entry point of the taint.
    - The `to` metavariable specifies where the tainted data is propagated to, typically an object or data structure.
```

```

In the example above, whenever Semgrep finds the pattern `$X.add($Y)`, it checks if
`$Y` is tainted. If `$Y` is tainted, Semgrep propagates that taint to `$X`. 
Thus, adding tainted data to a set marks the set itself as tainted.

<iframe src="https://semgrep.dev/embed/editor?snippet=7lNe" border="0" frameBorder="0" width="100%" height="432"></iframe>

## Requirements of taint propagator rules

A taint propagator rule must contain:
- A `pattern` containing **two** metavariables. These two metavariables specify where taint is propagated **from** and **to**. For example, pattern `$X.add($Y)` includes two metavariables `$X` and `$Y`. Then specify which metavariable falls under `from` and which falls under `to` (see more information below).

> - Use `to` and `from` metavariables. Both of these metavariables must contain either a **variable** or an **expression**.

Author: ```suggestion
- The `to` and `from` metavariables. Both of these metavariables must contain either a **variable** or an **expression**.
```

```

In the example above, whenever Semgrep finds the pattern `$X.add($Y)`, it checks if
`$Y` is tainted. If `$Y` is tainted, Semgrep propagates that taint to `$X`. 
Thus, adding tainted data to a set marks the set itself as tainted.

<iframe src="https://semgrep.dev/embed/editor?snippet=7lNe" border="0" frameBorder="0" width="100%" height="432"></iframe>

## Requirements of taint propagator rules

A taint propagator rule must contain:
1. A `pattern` containing **two** metavariables. These two metavariables specify where taint is propagated **from** and **to**. For example, pattern `$X.add($Y)` includes two metavariables `$X` and `$Y`. Then specify which metavariable falls under `from` and which falls under `to` (see more information below).

> 2. The `to` and `from` metavariables. Both of these metavariables must contain either a **variable** or an **expression**.

Reviewer: ```suggestion
2. The `to` and `from` metavariables. These metavariables can contain either a **variable** or an **expression**.
```

Author: @efortuna thank you!
